src/gap_to_julia.jl

```diff
@@ -9,361 +9,424 @@
     else
-        throw(ConversionError(obj, Set{T}))
+        throw(ConversionError(obj, TT))
     end
-    len_list = length(obj)
+
+    recursive && recursion_dict !== nothing && haskey(recursion_dict, (obj, TT)) && return recursion_dict[(obj, TT)]
+    rec_dict = recursion_info_j(TT, obj, recursive, recursion_dict)
+
+    len_list = length(newobj)
     new_array = Vector{T}(undef, len_list)
-    if recursive
-        recursion_dict = IdDict()
-    end
     for i = 1:len_list
-        current_obj = ElmList(obj, i)
-        if recursive
-            new_array[i] = get!(recursion_dict, current_obj) do
-                gap_to_julia(T, current_obj, recursion_dict; recursive = true)
-            end
+        current_obj = ElmList(newobj, i)
+        if recursive && !isbitstype(typeof(current_obj))
+            new_array[i] =
+                gap_to_julia_internal(T, current_obj, rec_dict, Val(recursive))
         else
             new_array[i] = current_obj
         end
     end
-    return Set{T}(new_array)
+    ret_val = TT(new_array)
+    handle_recursion((obj, TT), ret_val, recursive, rec_dict)
+    return ret_val
 end
 
 ## Tuples
 ## Note that the tuple type prescribes the types of the entries,
 ## thus we have to convert at least also the next layer,
 ## even if `recursive == false` holds.
-function gap_to_julia(
-    ::Type{T},
+function gap_to_julia_internal(
+    TT::Type{T},
     obj::GapObj,
-    recursion_dict::RecDict = IdDict();
-    recursive::Bool = true,
-) where {T<:Tuple}
+    recursion_dict::JuliaCacheDict,
+    ::Val{recursive},
+) where {T<:Tuple, recursive}
+
     !Wrappers.IsList(obj) && throw(ConversionError(obj, T))
-    if !haskey(recursion_dict, obj)
-        parameters = T.parameters
-        len = length(parameters)
-        length(obj) == len ||
-            throw(ArgumentError("length of $obj does not match type $T"))
-        list = [
-            gap_to_julia(parameters[i], obj[i], recursion_dict; recursive)
-            for i = 1:len
-        ]
-        recursion_dict[obj] = T(list)
-    end
-    return recursion_dict[obj]::T
-end
 
-## Ranges
-gap_to_julia(::Type{T}, obj::GapObj; recursive::Bool = true) where {T<:UnitRange} = T(obj)
-gap_to_julia(::Type{T}, obj::GapObj; recursive::Bool = true) where {T<:StepRange} = T(obj)
+    recursive && recursion_dict !== nothing && haskey(recursion_dict, (obj, TT)) && return recursion_dict[(obj, TT)]
+    rec_dict = recursion_info_j(TT, obj, recursive, recursion_dict)
+
+    # extract the Tuple parameters, i.e. from Tuple{T1, T2, ...}  the list T1,T2,...
+    parameters = T.parameters
+    len = length(parameters)
+    length(obj) == len ||
+        throw(ArgumentError("length of $obj does not match type $T"))
+    list = [
+        gap_to_julia_internal(parameters[i], obj[i], rec_dict, Val(recursive))
+        for i = 1:len
+    ]
+
+    ret_val = T(list)
+    recursion_dict = handle_recursion((obj, TT), ret_val, recursive, rec_dict)
```

Comment:
Here in principle I have the same concern as for the `Set` case but at the same time it is just not possible to construct the Tuple *before* recursion, but it also shouldn't be necessary due to how identity works for tuples. I.e. if we construct the tuple multiple times, that's fine, only downside is that we also end up storing it multiple times.

Might be good to add some test cases exercising this "recursive data structure" situations ... So e.g.
```
julia> v = [1, "x"]
2-element Vector{Any}:
 1
  "x"

julia> t = (2, v)
(2, Any[1, "x"])

julia> v[2] = t
(2, Any[1, (#= circular reference @-2 =#)])

julia> v
2-element Vector{Any}:
 1
  (2, Any[#= circular reference @-2 =#])
```

...